File: campus-exchange/mobile/lib/models/conversation.dart
class Conversation {
  final String conversationId;
  final String listingId;
  final String collegeId;
  final String buyerId;
  final String sellerId;
  final String lastMessage;
  final String? lastMessageSenderId;
  final DateTime lastMessageAt;
  final String? listingTitle;
  final double? listingPrice;
  final String? listingPhoto;
  final String? listingStatus;

  Conversation({
    required this.conversationId,
    required this.listingId,
    required this.collegeId,
    required this.buyerId,
    required this.sellerId,
    required this.lastMessage,
    this.lastMessageSenderId,
    required this.lastMessageAt,
    this.listingTitle,
    this.listingPrice,
    this.listingPhoto,
    this.listingStatus,
  });

  factory Conversation.fromJson(Map<String, dynamic> json) {
    return Conversation(
      conversationId: json['conversationId'] ?? '',
      listingId: json['listingId'] ?? '',
      collegeId: json['collegeId'] ?? '',
      buyerId: json['buyerId'] ?? '',
      sellerId: json['sellerId'] ?? '',
      lastMessage: json['lastMessage'] ?? '',
      lastMessageSenderId: json['lastMessageSenderId'],
      lastMessageAt: json['lastMessageAt'] != null
          ? DateTime.tryParse(json['lastMessageAt']) ?? DateTime.now()
          : DateTime.now(),
      listingTitle: json['listingTitle'],
      listingPrice: json['listingPrice'] != null ? (json['listingPrice'] as num).toDouble() : null,
      listingPhoto: json['listingPhoto'],
      listingStatus: json['listingStatus'],
    );
  }
}

File: server/src/models/Conversation.js
const mongoose = require('mongoose');

const ConversationSchema = new mongoose.Schema(
  {
    conversationId: {
      type: String,
      required: true,
      unique: true,
      index: true,
    },
    listingId: {
      type: String,
      required: true,
      index: true,
      ref: 'Listing',
    },
    collegeId: {
      type: String,
      required: true,
      index: true,
      ref: 'College',
    },
    buyerId: {
      type: String,
      required: true,
      index: true,
      ref: 'Student',
    },
    sellerId: {
      type: String,
      required: true,
      index: true,
      ref: 'Student',
    },
    lastMessage: {
      type: String,
      default: '',
    },
    lastMessageSenderId: {
      type: String,
    },
    lastMessageAt: {
      type: Date,
      default: Date.now,
      index: true,
    },
  },
  {
    timestamps: true,
  }
);

// One conversation per listing + buyer combination
ConversationSchema.index({ listingId: 1, buyerId: 1 }, { unique: true });
ConversationSchema.index({ collegeId: 1, buyerId: 1, lastMessageAt: -1 });
ConversationSchema.index({ collegeId: 1, sellerId: 1, lastMessageAt: -1 });

module.exports = mongoose.model('Conversation', ConversationSchema);
